### PR TITLE
Allow users to embed data a Request

### PR DESCRIPTION
**Pull request by davidbirdsong**
I need to attach some data to every request, would you consider a small [change to the  Request interface](https://github.com/davidbirdsong/vulcan/commit/3dbdf89308f852f30e319e50d119a280a963cb74)?

### CODE REVIEW

File: request/request.go [hunk 1]
```diff
@@ -92,3 +108,22 @@ func (br *BaseRequest) GetLastAttempt() Attempt {
...
+	defer br.userDataMutex.Unlock()
+	if br.userData == nil {
+		br.userData = make(map[string]interface{})
+	}
+	br.userData[key] = baton
+}
+func (br *BaseRequest) GetUserData(key string) (i interface{}, b bool) {
+	br.userDataMutex.RLock()
+	defer br.userDataMutex.RUnlock()
+	i, b = br.userData[key]
```
Comment (klizhentas) [inline request/request.go:125]: This operation will crash if you haven't executed `SetUserData` before executing `GetUserData` as `userData` will bi nil, right?

**Comment by davidbirdsong** [comment]
Good catch--just pushed up a change.

I'm opting for not calling `make` unless the user wants to actually store something since a Request object has a high rate of creation.

**Comment by klizhentas** [comment]
Thanks! One last question - can you please add a specific test case that covers Get without prior Set so we have this behavior enforced in test.

Otherwise looks good!

**Comment by davidbirdsong** [comment]
Yep, good call. I've added that test.

**Pull request merged**